src/models.js

```diff
@@ -108,22 +113,68 @@ function renderLiquid (world, cursor, texture, type, water, attr) {
   }
 }
 
-function renderElement (world, cursor, element, doAO, attr) {
+function vecadd3 (a, b) {
+  if (!b) { return a }
+  return a.map((v, i) => v + b[i])
+}
+
+function vecsub3 (a, b) {
+  if (!b) { return a }
+  return a.map((v, i) => v - b[i])
+}
+
+function matmul3 (matrix, vector) {
+  if (!matrix) { return vector }
+
+  const result = [0, 0, 0]
+  for (let i = 0; i < 3; ++i) {
+    for (let j = 0; j < 3; ++j) { result[i] += matrix[i][j] * vector[j] }
+  }
+  return result
+}
+
+function buildRotationMatrix (axis, degree) {
+  const radians = degree / 180 * Math.PI
+  const cos = Math.cos(radians)
+  const sin = Math.sin(radians)
+
+  const axis0 = { x: 0, y: 1, z: 2 }[axis]
+  const axis1 = (axis0 + 1) % 3
+  const axis2 = (axis0 + 2) % 3
+
+  const matrix = [
+    [0, 0, 0],
+    [0, 0, 0],
+    [0, 0, 0]
+  ]
+
+  matrix[axis0][axis0] = 1
+  matrix[axis1][axis1] = cos
+  matrix[axis1][axis2] = -sin
+  matrix[axis2][axis1] = +sin
+  matrix[axis2][axis2] = cos
+
+  return matrix
+}
+
+function renderElement (world, cursor, element, doAO, attr, globalMatrix, globalShift, block) {
   for (const face in element.faces) {
     const eFace = element.faces[face]
     const { dir, corners, mask1, mask2 } = elemFaces[face]
 
     if (eFace.cullface) {
       const neighbor = world.getBlock(cursor.plus(dir))
-      if (!neighbor || !(neighbor.transparent || !neighbor.isCube) || neighbor.position.y < 0) continue
```

Comment:
If you want to address the culling issue later, you can revert this change so we can merge already the other improvements.

Reply:
Thanks for letting me know. I have fixed this in d502c90a740cbad3e476896b086da71c327f5919. Apparently this was also the reason the section processing time went up, which is now back to what it was before this PR (i.e. doing the rotations on the client does not cause any noticeable overhead).